src/rp2/tax_engine.py

```diff
@@ -194,7 +194,10 @@ def _create_unfiltered_gain_and_loss_set(
                 )
 
     except AcquiredLotsExhaustedException:
-        raise RP2ValueError("Total in-transaction crypto value < total taxable crypto value") from None
+        raise RP2ValueError(
```

Comment:
Unfortunately this is not quite correct: line 196 doesn't capture `acquired_lot_amount - taxable_event_amount < 0` (this is already captured above at line 179 and is a condition related to an individual couple of acquired lot and taxable event). Line 196 captures the fact that acquired lots have been exhausted for the given taxable event using the given accounting method, so it's not even possible to try with the next acquired lot, because there are no more (unlike what happens at line 176).

When running into this, the solution is to enable debug logs (by prepending the command with `LOG_LEVEL=DEBUG`) and carefully analyze the various logs starting with "tax_engine: " (see above in this file). This will lead to pinpointing the problem. 

I think this is an important issue though: it probably deserves a user FAQ.

Reply:
Thanks for the feedback.  I will look into this further (run through debug logs for a specific case) when I have a chance.

It might be good to give some user direction in the exception output, even if that is a redirection to LOG_LEVEL=DEBUG.  In my case I did find the time and value at the time of occurrence useful, but it sounds like I at least got the value wrong.

Reply:
Yes, I think this PR could change to:
* new user FAQ with the description of the problem and the procedure to solve it;
* add a pointer to the new FAQ to the error message.

Reply:
Please consider the following and let me know your thoughts.

Setting an environment variable varies by environment.  In PowerShell for example, it needs to be set via a different execution line, such as:
`$Env:LOG_LEVEL = "DEBUG"`

I modified my data set to reintroduce one of the issues I cited and ran with LOG_LEVEL=DEBUG.  If we are going to redirect the user to debug logging, they will need to navigate to the end of the log file and look for the DEBUG: tax_engine as you suggest, where the last entry will be taxable > acquired.  It is likely there will be many other entries with id=y->x as it searches for acquired lots to meet the taxable out transaction.  In my case, there were almost 300 lots (x) to try and account for y.  There is also no direct output to the user the amount they are negative by; the user needs to calculate it manually from the last "DEBUG: tax_engine: taxable > acquired ..." output.

The change I made does output the transaction information for the correct id (y).  I think it does well to point the user in the right direction.  Each time I had this problem I needed to know when/where I went negative and scrutinize in that area, and this clearly points that out.

I recommend still printing the offending output transaction information when this condition is met, since it is a fatal exception and I think most of the time this will be enough information for the user to determine what the problem is and they will not actually need to turn on debug.  I think this is a more user friendly approach and that is goal, to make transitioning to DaLI and RP2 easier.

Whether you want the output transaction in the normal output or not, I can take a stab at modifying the output to refer to an FAQ section where both DEBUG and review of the DEBUG logs can be described.

Reply:
Looking at this a bit more it does seem the values I am populating for "Total in-transaction crypto value - total taxable crypto value" are for the individual taxable and acquired lot portion, and not the total.  Maybe something like:

```
Taxable out-transaction crypto value > acquired in-transaction crypto value by 0.012345 for OutTransaction:
  id=y
  ...

For more information, see Some Section in the User FAQ: https://github.com/eprbell/rp2/blob/main/docs/user_faq.md#somenewsection
```

Reply:
Thanks for looking into this: it's indeed one of the least user friendly errors that can occur in RP2 and it's good to think about how to improve it. I'm not quite sure I understand your proposal: could you upload the code so that we can brainstorm on it? 

Your message raises some good points: it's not great to ask users to resort to DEBUG logging (that should be reserved for developers). Here's the issue as I see it: if execution completes successfully, the user has the rp2_full_report.ods file, which contains the full history of in/out lot mappings. However if this error occurs, the user doesn't have that history, which would actually be very helpful in debugging the error.

What if we captured the evolution of the `_create_unfiltered_gain_and_loss_set()` function? We could write to the `output` directory a csv file that has one row per GainLoss object. This way if anything goes wrong the user has all the information handy and we could point them to the file.

Reply:
When I ran into this problem I sorted crypto_data.ods so ins and outs and intras were in chronological order, and then I used a formula to calculate balance at each transaction.  This gave me a look no further past this point to start with.  Luckily, the source of the problem wasn't much earlier than that, but depending on the user's situation it could be all the way at the first in transaction if that is where the data is bad.

Thinking a bit more, I've run into this same problem using the proprietary options and they sometimes provide the audit history if they have a way to view specific tax lots.  Generally they tell you which transaction has the problem and flag it, and it is up to the user to figure out what the problem is.  They do provide help information.  Some examples:
https://support.cointracker.io/hc/en-us/articles/22620891108881-How-to-Resolve-Transactions-With-Missing-Balance
https://help.blockpit.io/hc/en-us/articles/360012172479-What-do-Tips-mean-and-how-to-solve-them#zp-2-2

The proprietary options also find a way to carry on with the remainder of the calculations but flag it as an error that the user needs to fix.  The user can make the auto-balance permanent.  One option would be to add an auto-balance command line option which adds the missing amounts so everything can generate and then the user would have the full data set to work with in the normal output.  In this flow RP2 would fail the first time, they would add the option, get a generated data set, and use that to investigate further.
https://help.blockpit.io/hc/en-us/articles/360011984959-Transaktions-Label-Auto-Balancing

Could also consider adding a balance column to each transaction in the In-Out sheet (Ledger View).  This would be similar to the running sums but would take account ins, outs, and intras.
https://help.blockpit.io/hc/en-us/articles/8814643893020-Advanced-Transaction-Details-Ledger-Tax-View

I think with this approach we can get the user the information they need without them ever needing to turn debug on.

Reply:
Is crypto_data.ods your input file? Sorting it chronologically only works with FIFO, not with LIFO or HIFO. Also adding a balance column to the output file doesn't help in this case, because if the error happens there is no output file to look at (but it's something we could consider doing as a separate enhancement). This is why I think we should generate a temporary file with the GainLossSet history: this would work even if the output file is not generated and it would reflect the order mandated by the accounting method (FIFO, LIFO or HIFO). Essentially the temporary file contains information equivalent to the "Gain / Loss Detail" table in the rp2_full_report.ods output file (up to the error). We could also add the balance column to the temporary file.

Reply:
Yes, my input file.  I was running RP2 with both FIFO and HIFO to see how they compare, and if you are going negative it doesn't matter which one you use any ways.  If everything is out, it doesn't matter which order it went, you will get to zero and then negative eventually.

You are right, having the balance listed would only be useful if things could actually generate, which is possible by creating a balancing transaction, either by RP2 or manually.  In my specific cases, the missed transaction simply had to be added, the typo in a transaction simply had to be fixed, but for the rounding errors (problem introduced by the exchange, not me) I actually changed my data set to reduce the crypto out value to balance at exactly zero.  That's the opposite of auto-balancing, which usually adds an in transaction.

Any ways, I'll see about finding some time to take a stab at some code changes.

Reply:
My point is that sorting chronologically is an approximation of what FIFO does (e.g. it doesn't capture errors like: current in lot acquired after current out lot). Additionally when HIFO and LIFO is used fraction pairings will be different. So it may work in certain situations but it's not something we can recommend to the users. The temporary CSV file would solve all the above problems and accurately reflect what the tax engine is computing.

Reply:
Converting to CSV is being a pain, but I am making progress.  Some questions for you so I don't have to keep wrestling with CSV columns.

Do you want to output all gain losses for this particular asset or only gain losses for the last taxable event?  In the later, if the first acquired lot in the last taxable event processed is a partial, the user will not see any lots from previous taxable events.  crypto_amount will be important in either case, but in this one it will be less than crypto_in.

What fields do you want in the CSV?  There are almost 40 fields if we put them all in, although OutTransaction will be unnecessary if we are only exporting the last taxable event.

Reply:
The way I see it is that the CSV files are meant as an "in progress" version of the "Gain / Loss Detail" table in the rp2_full_report.ods output file. So they should contain similar data. We could have one CSV file per asset to keep things easy for the user.

I suggest appending one new CSV row for the current gain loss entry at this location: https://github.com/eprbell/rp2/blob/main/src/rp2/gain_loss_set.py#L129 (just before the "if" statement). 

For reference you can look at `_fill_cell` calls here: https://github.com/eprbell/rp2/blob/main/src/rp2/plugin/report/rp2_full_report.py#L802 (it's where the columns of the "Gain / Loss Detail" table are filled).

Reply:
Thanks for the pointers, I certainly don't know the best place in the code to insert new behavior.

I was not suggesting one CSV per asset.  One asset should not have direct impact on the other.  Also, since this error condition results in termination, not all assets will be processed.

If the intent is to show the user as much information as possible (all assets, not just the one that ran into the error), and the intent is to show the user information in a manner similar to the Asset Tax sheet in the full report, I wonder if it is just best to ignore the difference and keep going, and reporting it to the user (something like an auto-balance I was suggesting).  They will know there is an issue to look into, they will get the output in a similar manner as they would the full report, and they will get all of the other features like being able to click around and navigate throughout the ODS report.  No need to use CSV files with no "features".

I think in either case it is beneficial to add a balance column to the In-Out sheet and now I guess the Tax sheet as well, since Crypto (In/Out/Intra/Amt) Running Sum is in both of those.

Thoughts?

Reply:
Interesting thought... this will need to be experimented with a bit. We need to understand if there are any side effects of not interrupting execution. I guess to start with we can change the exceptions that are raised in `tax_engine.py:_create_unfiltered_gain_and_loss_set()` to a new custom exception (e.g. RP2AccountingMethodException) and capture that in the caller, which is `compute_tax()` in the same file.

Let's do the balance column addition in a separate PR to keep things simple (that will require updating the golden files).

Reply:
OK, I will work on the balance column addition first, probably not this week.

Reply:
I have a working prototype for in, out, and intra balance columns.  Gain/loss is not being calculated correctly.  I think once I get that working I might ask for an initial review before I start updating automated testing, gold, etc.  More work than I expected and it is now the nice weather months so this might take a bit!